src/utils/metrics/metrics.go

```diff
@@ -20,113 +20,289 @@
-	return &Writer{ms: ms, jobID: jobID, name: name, value: 0}
+	w := tabwriter.NewWriter(target, 1, 1, 1, ' ', tabwriter.AlignRight)
+
+	defer w.Flush()
+
+	stats := r.SumAllStatsByTarget()
+	if stats.sum(RequestsSentStat) == 0 {
+		fmt.Fprintln(w, "[Error] No traffic generated. If you see this message a lot - contact admins")
+
+		return
+	}
+
+	fmt.Fprintln(w, "\n\n!Атака проводиться успішно! Русскій воєнний корабль іди нахуй!")
+	fmt.Fprintln(w, "!Attack is successful! Russian warship, go fuck yourself!")
+
+	fmt.Fprintln(w, "\n --- Traffic stats ---")
+	fmt.Fprintf(w, "|\tTarget\t|\tRequests attempted\t|\tRequests sent\t|\tResponses received\t|\tData sent \t|\n")
+
+	const BytesInMegabyte = 1024 * 1024
+
+	var totals Stats
+
+	for _, tgt := range stats.sortedTargets() {
+		tgtStats := stats[tgt]
+
+		fmt.Fprintf(w, "|\t%s\t|\t%d\t|\t%d\t|\t%d\t|\t%.2f MB \t|\n", tgt,
+			tgtStats[RequestsAttemptedStat],
+			tgtStats[RequestsSentStat],
+			tgtStats[ResponsesReceivedStat],
+			float64(tgtStats[BytesSentStat])/BytesInMegabyte,
+		)
+
+		for s := range totals {
+			totals[s] += tgtStats[s]
+		}
+	}
+
+	fmt.Fprintln(w, "|\t---\t|\t---\t|\t---\t|\t---\t|\t--- \t|")
+	fmt.Fprintf(w, "|\tTotal\t|\t%d\t|\t%d\t|\t%d\t|\t%.2f MB \t|\n\n",
+		totals[RequestsAttemptedStat],
+		totals[RequestsSentStat],
+		totals[ResponsesReceivedStat],
+		float64(totals[BytesSentStat])/BytesInMegabyte,
+	)
 }
 
-// Writer is a helper to accumulate writes to a storage on a regular basis
-type Writer struct {
-	ms    *Storage
-	jobID string
-	name  string
-	value uint64
+// NewAccumulator returns a new metrics Accumulator for the Reporter.
+func (r *Reporter) NewAccumulator(jobID string) *Accumulator {
+	if r == nil {
+		return nil
+	}
+
+	return newAccumulator(jobID, r)
 }
 
-// Add used to increase metric value by a specific amount
-func (w *Writer) Add(value uint64) {
-	w.value += value
+// Accumulator for statistical metrics for use in a single job. Requires Flush()-ing to Reporter.
+// Not concurrency-safe.
+type Accumulator struct {
+	jobID   string
+	metrics [NumStats]map[string]uint64 // Array of metrics by Stat. Each metric is a map of uint64 values by target.
+
+	r *Reporter
 }
 
-// Set used to set metric to a specific value
-func (w *Writer) Set(value uint64) {
-	w.value = value
+// Add n to the Accumulator Stat value. Returns self for chaining.
+func (a *Accumulator) Add(target string, s Stat, n uint64) *Accumulator {
+	if a == nil || !s.valid() {
+		return a
+	}
+
+	a.metrics[s][target] += n
+
+	return a
 }
 
-// Flush used to flush pending metrics updates to the storage
-func (w *Writer) Flush() {
-	w.ms.Write(w.name, w.jobID, w.value)
+// Inc increases Accumulator Stat value by 1. Returns self for chaining.
+func (a *Accumulator) Inc(target string, s Stat) *Accumulator { return a.Add(target, s, 1) }
+
+// AddStats to the Accumulator. Returns self for chaining.
+func (a *Accumulator) AddStats(target string, s Stats) *Accumulator {
```

Comment:
Can we have a function that would expect a variadic list of pairs <Stat, value> instead? The way this API is implemented now it's easy to misuse it by passing a slice of wrong size or reordering arguments accidentally

Reply:
See the `NewStats` constructor instead which should solve your concern